File: 260_single_number/single_number.go
/*
 * @Date: 2021-10-30 01:13:14
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-30 01:25:02
 */

package main

import "reflect"

func singleNumber(nums []int) []int {
	xorSum := 0
	for _, num := range nums {
		xorSum ^= num
	}
	lsb := xorSum & -xorSum
	type1, type2 := 0, 0
	for _, num := range nums {
		if num&lsb > 0 {
			type1 ^= num
		} else {
			type2 ^= num
		}
	}
	return []int{type1, type2}
}

func main() {
	assert := func(a, b []int) {
		if !reflect.DeepEqual(a, b) {
			panic("NotPassed")
		}
	}
	assert(singleNumber([]int{1, 2, 1, 3, 2, 5}), []int{3, 5})
	assert(singleNumber([]int{-1, 0}), []int{-1, 0})
	assert(singleNumber([]int{0, 1}), []int{1, 0})
}

File: rust/260_single_number/single_number.rs
/*
 * @Date: 2021-10-30 01:13:07
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-10-16
 */

struct Solution;
impl Solution {
    pub fn single_number(nums: Vec<i32>) -> Vec<i32> {
        let mut all = 0;
        for num in &nums {
            all ^= num;
        }
        let diff = all & (-all);
        let mut x = 0;
        for num in &nums {
            if (diff & num) != 0 {
                x ^= num;
            }
        }
        vec![x, x ^ all]
    }
}

fn main() {
    let tests = vec![
        (vec![1, 2, 1, 3, 2, 5], vec![3, 5]),
        (vec![-1, 0], vec![-1, 0]),
        (vec![0, 1], vec![1, 0]),
    ];

    for (nums, ans) in tests {
        assert_eq!(Solution::single_number(nums), ans);
    }
}
